src/Microsoft.Maintainability.Analyzers/Core/ReviewUnusedParameters.cs

```diff
@@ -23,15 +28,131 @@ public abstract class ReviewUnusedParametersAnalyzer : DiagnosticAnalyzer
+
+                var attributeSetForMethodsToIgnore = ImmutableHashSet.Create(conditionalAttributeSymbol, onDeserializedAttribute, onDeserializingAttribute, onSerializedAttribute, onSerializingAttribute);
+
+                compilationStartContext.RegisterOperationBlockStartAction(startOperationBlockContext =>
+                {
+                    // We only care about methods.
+                    if (startOperationBlockContext.OwningSymbol.Kind != SymbolKind.Method)
+                    {
+                        return;
+                    }
+
+                    // We only care about methods with parameters.
+                    var method = (IMethodSymbol)startOperationBlockContext.OwningSymbol;
+                    if (method.Parameters.IsEmpty)
+                    {
+                        return;
+                    }
+
+                    // Ignore implicitly declared methods, abstract methods, virtual methods, interface implementations and finalizers (FxCop compat).
+                    if (method.IsImplicitlyDeclared ||
+                        method.IsAbstract ||
+                        method.IsVirtual ||
+                        method.IsOverride ||
+                        method.IsImplementationOfAnyInterfaceMethod() ||
+                        method.IsFinalizer())
+                    {
+                        return;
+                    }
+
+                    // Ignore event handler methods "Handler(object, MyEventArgs)"
+                    if (eventsArgSymbol != null &&
+                        method.Parameters.Length == 2 &&
+                        method.Parameters[0].Type.SpecialType == SpecialType.System_Object &&
+                        method.Parameters[1].Type.Inherits(eventsArgSymbol))
+                    {
+                        return;
+                    }
+
+                    // Ignore methods with any attributes in 'attributeSetForMethodsToIgnore'.
+                    if (method.GetAttributes().Any(a => a.AttributeClass != null && attributeSetForMethodsToIgnore.Contains(a.AttributeClass)))
+                    {
+                        return;
+                    }
+
+                    // Initialize local mutable state in the start action.
+                    var analyzer = new UnusedParametersAnalyzer(method);
+
+                    // Register an intermediate non-end action that accesses and modifies the state.
+                    startOperationBlockContext.RegisterOperationAction(analyzer.AnalyzeOperation, OperationKind.ParameterReferenceExpression);
+
+                    // Register an end action to report diagnostics based on the final state.
+                    startOperationBlockContext.RegisterOperationBlockEndAction(analyzer.OperationBlockEndAction);
+                });
+            });
+        }
+
+        private class UnusedParametersAnalyzer
+        {
+            #region Per-CodeBlock mutable state
```

Comment:
rather than you implement this yourself, you might just use this.

DataFlowAnalysis - http://source.roslyn.io/#Microsoft.CodeAnalysis/Compilation/DataFlowAnalysis.cs,bfbcd96bb21963de
from semantic model.

you can check ReadInside and WrittenInside to find out parameters that are used in the scope (here scope will be method body)